### PR TITLE
#11569 - Add duplicate detection to import

### LINKED ISSUE #11569: Add duplicate detection for environment import [3]
### Feature Description

This issue is part of the EPIC https://github.com/hzi-braunschweig/SORMAS-Project/issues/11533 on the environmental surveillance module. Add a duplicate detection for environments that triggers when environments are imported through a .csv file.

The following fields MUST be filled in in order for two environments being considered as duplicates:
- Environment media (must match)
- GPS lat (the two GPS lat coordinates must be within 0.02 digits from each other; i.e. 52.288 would be considered a potential duplicate for GPS lat values between 52.286 and 52.290; this is roughly 0.5 kilometers)
- GPS lon (same as GPS lat)

Additionally, if the following fields are specified in BOTH environments, they must match:
- Country
- Region
- District

If an external ID is specified for both environments, they MUST match and if they do, they overwrite all the other rules (i.e. two environments with the same external ID are considered as potential duplicates even if they're in different regions).

### Acceptance Criteria

- [x] The environment import window has been extended by a figure that lists the number of environments that have been skipped as duplicates (infrastructure import should already have something like this)
- [x] When importing environments, the environment is checked for potential duplicates in the database; please note that this has to happen in the database, without querying for all other environments first
- [x] If at least one potential duplicate has been found, the environment is not imported, but counted as a skip and written into the error log with the following info text in the first column: "Skipped as a potential duplicate"
- [x] An error log is produced and downloadable if at least one environment has been skipped, even if there were no other import errors
- [ ] The import guide has been extended with information about downloading the error log when duplicates have been detected (can be done by @MateStrysewske)

### Additional Information

Duplicate detection and merging (as we have it for cases) of Environments at Creation will be done in a later phase.
This duplicate detection is limited to import environments.

### PR DESCRIPTION
**Pull request by MateStrysewske**
Fixes #11569

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12554

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12554

File: sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentService.java [hunk 1]
```diff
@@ -249,6 +267,83 @@ public Predicate buildCriteriaFilter(EnvironmentCriteria environmentCriteria, En
...
+				cb.isNotNull(joins.getLocation().get(Location.LATITUDE)),
+				cb.greaterThanOrEqualTo(joins.getLocation().get(Location.LATITUDE), criteria.getGpsLatFrom() - 0.02),
+				cb.lessThanOrEqualTo(joins.getLocation().get(Location.LATITUDE), criteria.getGpsLatTo() + 0.02)));
+		filter = CriteriaBuilderHelper.and(
+			cb,
+			filter,
+			cb.and(
+				cb.isNotNull(joins.getLocation().get(Location.LONGITUDE)),
+				cb.greaterThanOrEqualTo(joins.getLocation().get(Location.LONGITUDE), criteria.getGpsLonFrom() - 0.02),
+				cb.lessThanOrEqualTo(joins.getLocation().get(Location.LONGITUDE), criteria.getGpsLonTo() + 0.02)));
```
Comment (sergiupacurariu) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/environment/EnvironmentService.java:311]: GPS latitude and longitude can take also negative values in this case the evaluation of "-0.02" and "+0.02" should work the other way around. Also, there is a situation when you are within less than 0.02 from the equator or from the prime meridian.

File: sormas-backend/src/test/java/de/symeda/sormas/backend/environment/EnvironmentServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+public class EnvironmentServiceTest extends AbstractBeanTest {
+
+	@Test
+	public void testGetSimilarEnvironmentUuid() {
+
+		RDCF rdcf = creator.createRDCF();
+		RDCF rdcf2 = creator.createRDCF("Region2", "District2", "Community2", "Facility2");
+
+		EnvironmentDto environment1 = creator.createEnvironment("Environment1", EnvironmentMedia.AIR, nationalAdmin.toReference(), rdcf, e -> {
+			e.getLocation().setLatitude(20D);
```
Comment (sergiupacurariu) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/environment/EnvironmentServiceTest.java:38]: the GPS Latitude values should have a range between [-90, 90]

File: sormas-backend/src/test/java/de/symeda/sormas/backend/environment/EnvironmentServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+			e.getLocation().setLongitude(20D);
```
Comment (sergiupacurariu) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/environment/EnvironmentServiceTest.java:39]: the GPS Longitude values should have a range between [-180, 180]

**Review by sergiupacurariu — CHANGES_REQUESTED** [review]

**Comment by sergiupacurariu** [comment]
when an Environment that is imported has all the fields that would match a duplicate but does not have an environment name which is mandatory, the error is that there is a duplicate instead of a missing mandatory field.

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12554

**Review by sergiupacurariu — APPROVED** [review]

**Pull request merged**